compiler/rustc_lint/src/nonstandard_style.rs

```diff
@@ -489,22 +492,82 @@ declare_lint! {
 declare_lint_pass!(NonUpperCaseGlobals => [NON_UPPER_CASE_GLOBALS]);
 
 impl NonUpperCaseGlobals {
-    fn check_upper_case(cx: &LateContext<'_>, sort: &str, ident: &Ident) {
+    fn check_upper_case(cx: &LateContext<'_>, sort: &str, did: Option<LocalDefId>, ident: &Ident) {
         let name = ident.name.as_str();
         if name.chars().any(|c| c.is_lowercase()) {
             let uc = NonSnakeCase::to_snake_case(name).to_uppercase();
+
+            // If the item is exported, suggesting changing it's name would be breaking-change
+            // and could break users without a "nice" applicable fix, so let's avoid it.
+            let can_change_usages = if let Some(did) = did {
+                !cx.tcx.effective_visibilities(()).is_exported(did)
+            } else {
+                false
+            };
+
             // We cannot provide meaningful suggestions
             // if the characters are in the category of "Lowercase Letter".
             let sub = if *name != uc {
```

Comment:
Could you move `let can_change_usages` and `let sub` into the callback of `emit_span_lint_lazy`? Should fix perf

Reply:
I could but I don't actually think there is any regression here, the only regressed crate is clap_derive-4.5.3/opt/full, no other benchmark, including any of the others clap_derive, or the previously regressed libc crate seems to affected and looking at the details the regression is in the backend, not the frontend. I think it's a spurious regression. What do you think?